examples/cookbook.cpp

```diff
@@ -0,0 +1,951 @@
+#include <iostream>
+
+#include <js/Conversions.h>
+#include <js/Initialization.h>
+#include <jsapi.h>
+#include <mozilla/Unused.h>
+
+/* This example program shows the SpiderMonkey JSAPI equivalent for a handful of
+ *common JavaScript idioms. */
+
+/**** BASICS ******************************************************************/
+
+///// Working with Values //////////////////////////////////////////////////////
+
+/* The basic, undifferentiated value type in the JSAPI is `JS::Value`.
+ * To query whether a value has a particular type, use a correspondingly named
+ * member testing function:
+ *
+ * // JavaScript
+ * var isString = typeof v === "string";
+ * var isNumber = typeof v === "number";
+ * var isNull = v === null;
+ * var isBoolean = typeof v === "boolean";
+ * var isObject = typeof v === "object" && v !== null;
+ */
+static void
+GetTypeOfValue(JSContext* cx, JS::HandleValue v)
+{
+  bool isString = v.isString();
+  bool isNumber = v.isNumber();
+  bool isInt32 =
+    v.isInt32(); // NOTE: Internal representation, not numeric value
+  bool isNull = v.isNull();
+  bool isBoolean = v.isBoolean();
+  bool isObject =
+    v.isObject(); // NOTE: not broken like typeof === "object" is :-)
+
+  // Avoid compiler warnings
+  mozilla::Unused << isString;
+  mozilla::Unused << isNumber;
+  mozilla::Unused << isInt32;
+  mozilla::Unused << isNull;
+  mozilla::Unused << isBoolean;
+  mozilla::Unused << isObject;
+}
+
+/* To set a value use a correspondingly named member mutator function, or assign
+ * the result of the correspondingly named standalone function:
+ *
+ * // JavaScript
+ * var v;
+ *
+ * v = 0;
+ * v = 0.5;
+ * v = someString;
+ * v = null;
+ * v = undefined;
+ * v = false;
+ */
+static void
+SetValue(JSContext* cx)
+{
+  JS::RootedValue v(cx);
+  JS::RootedString someString(cx, JS_NewStringCopyZ(cx, "my string"));
+
+  // clang-format off
+  v.setInt32(0);           /* or: */ v = JS::Int32Value(0);
+  v.setDouble(0.5);        /* or: */ v = JS::DoubleValue(0.5);
+  v.setString(someString); /* or: */ v = JS::StringValue(someString);
+  v.setNull();             /* or: */ v = JS::NullValue();
+  v.setUndefined();        /* or: */ v = JS::UndefinedValue();
+  v.setBoolean(false);     /* or: */ v = JS::BooleanValue(false);
+  // clang-format on
+}
+
+///// Finding the global object ////////////////////////////////////////////////
```

Comment:
The comments here are a little confusing. None of the recipes below actually use this helper so this doesn't really motivate why someone would need this. I'm not sure what best way to explain this without getting into the weeds about wrappers and why finding the global is complicated.

I'm fine to leave this as is, but it might need a future cleanup.

Reply:
I don't know why finding the global object is complicated myself, but I've tried to make the comments less confusing at least.